File: Tutorials/Codes/Modern_Control_Course/Chapter7/Lesson3/Chapter7_Lesson3.py
"""
Chapter7_Lesson3.py
Modern Control — Chapter 7 (Solutions of LTI State Equations), Lesson 3
Solutions for Constant, Step, and Polynomial Inputs

This script demonstrates:
  1) Constant input u(t)=u0
  2) Step input u(t)=u0 * 1(t-ts)
  3) Polynomial input u(t)=u0 + u1 t + u2 t^2

We compute closed-form expressions using matrix exponentials and phi-functions.
"""

import numpy as np

try:
    from scipy.linalg import expm
    from scipy.integrate import solve_ivp
except ImportError as e:
    raise SystemExit("This script requires SciPy (scipy.linalg.expm, scipy.integrate.solve_ivp).") from e


def phi_series(Z: np.ndarray, m: int, terms: int = 40) -> np.ndarray:
    """
    Compute phi_m(Z) = sum_{j=0}^∞ Z^j / (j+m)!  via truncated series.
    Works for any square Z (including singular). For larger norms, increase terms
    or use more sophisticated algorithms.
    """
    n = Z.shape[0]
    I = np.eye(n)
    # Start with j=0 term
    fact = math.factorial(m)
    S = I / fact
    Zpow = I.copy()
    for j in range(1, terms):
        Zpow = Zpow @ Z
        denom = math.factorial(j + m)
        S = S + Zpow / denom
    return S


def x_constant_input(A, B, x0, u0, t):
    """
    x(t) = e^{At} x0 + ( ∫_0^t e^{A(t-τ)} B dτ ) u0
         = e^{At} x0 + t * phi_1(A t) B u0
    """
    At = A * t
    E = expm(At)
    Phi1 = phi_series(At, m=1)
    return E @ x0 + (t * (Phi1 @ (B @ u0)))


def x_step_input(A, B, x0, u0, t, ts):
    """
    u(t) = u0 * 1(t-ts)
    x(t) = e^{At} x0                                  , t < ts
         = e^{At} x0 + ∫_{ts}^t e^{A(t-τ)} B u0 dτ     , t >= ts
         = e^{At} x0 + (t-ts) * phi_1(A (t-ts)) B u0   , t >= ts
    """
    if t < ts:
        return expm(A * t) @ x0
    else:
        E = expm(A * t)
        dt = t - ts
        Phi1 = phi_series(A * dt, m=1)
        return E @ x0 + (dt * (Phi1 @ (B @ u0)))


def x_poly_input(A, B, x0, coeffs, t):
    """
    Polynomial input: u(t) = sum_{k=0}^p u_k t^k, where u_k are vectors.
    Closed form:
      x(t) = e^{At} x0 + sum_{k=0}^p t^{k+1} k! * phi_{k+1}(A t) B u_k

    coeffs: list [u0, u1, ..., up], each shape (m,)
    """
    At = A * t
    E = expm(At)
    x = E @ x0
    for k, uk in enumerate(coeffs):
        Phik1 = phi_series(At, m=k+1)  # phi_{k+1}(At)
        x = x + (t ** (k + 1)) * math.factorial(k) * (Phik1 @ (B @ uk))
    return x


def simulate_ivp(A, B, x0, u_fun, t_span, t_eval):
    """
    Numerical verification by solving xdot = A x + B u(t)
    """
    A = np.asarray(A)
    B = np.asarray(B)
    x0 = np.asarray(x0)

    def f(t, x):
        return (A @ x) + (B @ u_fun(t))

    sol = solve_ivp(f, t_span=t_span, y0=x0, t_eval=t_eval, rtol=1e-10, atol=1e-12)
    return sol.t, sol.y.T


def main():
    # Example: 2-state, 1-input system
    A = np.array([[0.0, 1.0],
                  [-2.0, -3.0]])
    B = np.array([[0.0],
                  [1.0]])
    x0 = np.array([1.0, 0.0])

    u0 = np.array([2.0])  # constant / step amplitude

    # Time grid
    t_eval = np.linspace(0.0, 5.0, 251)

    # 1) Constant input
    x_cf = np.vstack([x_constant_input(A, B, x0, u0, t) for t in t_eval])

    t_num, x_num = simulate_ivp(A, B, x0, lambda t: u0, (t_eval[0], t_eval[-1]), t_eval)

    max_err = np.max(np.linalg.norm(x_cf - x_num, axis=1))
    print("Constant input: max ||x_closed - x_numeric|| =", max_err)

    # 2) Step input at ts
    ts = 1.5
    x_step_cf = np.vstack([x_step_input(A, B, x0, u0, t, ts) for t in t_eval])

    def u_step(t):
        return u0 if t >= ts else np.zeros_like(u0)

    t_num2, x_num2 = simulate_ivp(A, B, x0, u_step, (t_eval[0], t_eval[-1]), t_eval)

    max_err2 = np.max(np.linalg.norm(x_step_cf - x_num2, axis=1))
    print("Step input: max ||x_closed - x_numeric|| =", max_err2)

    # 3) Polynomial input u(t) = u0 + u1 t
    u1 = np.array([0.5])
    coeffs = [u0, u1]  # u(t)=u0 + u1 t
    x_poly_cf = np.vstack([x_poly_input(A, B, x0, coeffs, t) for t in t_eval])

    def u_poly(t):
        return u0 + u1 * t

    t_num3, x_num3 = simulate_ivp(A, B, x0, u_poly, (t_eval[0], t_eval[-1]), t_eval)

    max_err3 = np.max(np.linalg.norm(x_poly_cf - x_num3, axis=1))
    print("Polynomial input (affine): max ||x_closed - x_numeric|| =", max_err3)

    # Print a sample
    idx = 100
    print("\nSample at t =", t_eval[idx])
    print("x_const_closed =", x_cf[idx])
    print("x_step_closed  =", x_step_cf[idx])
    print("x_poly_closed  =", x_poly_cf[idx])


if __name__ == "__main__":
    import math
    main()

File: Tutorials/Codes/Modern_Control_Course/Chapter7/Lesson3/Chapter7_Lesson3.m
% Chapter7_Lesson3.m
% Modern Control — Chapter 7 (Solutions of LTI State Equations), Lesson 3
% Solutions for Constant, Step, and Polynomial Inputs
%
% This script demonstrates closed-form state responses using matrix exponentials
% and phi-functions for:
%   1) constant input u(t)=u0
%   2) step input u(t)=u0 * 1(t-ts)
%   3) polynomial input u(t)=u0 + u1 t
%
% It also verifies results by numerical integration via ode45.

clear; clc;

A = [0 1; -2 -3];
B = [0; 1];
x0 = [1; 0];

u0 = 2;
u1 = 0.5;

ts = 1.5;

tgrid = linspace(0, 5, 251);

% --- phi-function via truncated series: phi_m(Z) = sum_{j=0}^∞ Z^j/(j+m)! ---
phi_series = @(Z,m,terms) local_phi_series(Z,m,terms);

x_const = zeros(2, numel(tgrid));
x_step  = zeros(2, numel(tgrid));
x_poly  = zeros(2, numel(tgrid));

for i = 1:numel(tgrid)
    t = tgrid(i);

    % Constant input
    At = A*t;
    E = expm(At);
    Phi1 = phi_series(At, 1, 40);
    x_const(:,i) = E*x0 + t * (Phi1*(B*u0));

    % Step input
    if t < ts
        x_step(:,i) = expm(A*t)*x0;
    else
        dt = t - ts;
        Phi1_dt = phi_series(A*dt, 1, 40);
        x_step(:,i) = expm(A*t)*x0 + dt * (Phi1_dt*(B*u0));
    end

    % Polynomial input u(t) = u0 + u1 t
    % x(t) = e^{At}x0 + t*phi_1(At)B u0 + t^2 * 1! * phi_2(At)B u1
    Phi2 = phi_series(At, 2, 45);
    x_poly(:,i) = E*x0 + t * (Phi1*(B*u0)) + (t^2) * (Phi2*(B*u1));
end

% --- Numerical verification with ode45 ---
opts = odeset('RelTol',1e-10,'AbsTol',1e-12);

% 1) constant input
ode_const = @(t,x) A*x + B*u0;
[tn1, xn1] = ode45(ode_const, [tgrid(1) tgrid(end)], x0, opts);
xn1i = interp1(tn1, xn1, tgrid);

err1 = max(vecnorm((x_const' - xn1i), 2, 2));
fprintf('Constant input: max ||x_closed - x_numeric|| = %.3e\n', err1);

% 2) step input
ode_step = @(t,x) A*x + B*(u0*(t>=ts));
[tn2, xn2] = ode45(ode_step, [tgrid(1) tgrid(end)], x0, opts);
xn2i = interp1(tn2, xn2, tgrid);

err2 = max(vecnorm((x_step' - xn2i), 2, 2));
fprintf('Step input: max ||x_closed - x_numeric|| = %.3e\n', err2);

% 3) polynomial input
ode_poly = @(t,x) A*x + B*(u0 + u1*t);
[tn3, xn3] = ode45(ode_poly, [tgrid(1) tgrid(end)], x0, opts);
xn3i = interp1(tn3, xn3, tgrid);

err3 = max(vecnorm((x_poly' - xn3i), 2, 2));
fprintf('Polynomial input: max ||x_closed - x_numeric|| = %.3e\n', err3);

% Optional plot
figure;
plot(tgrid, x_const(1,:), tgrid, x_step(1,:), tgrid, x_poly(1,:));
xlabel('t'); ylabel('x_1(t)');
legend('constant', 'step', 'polynomial');
grid on;

% --- Simulink (optional): programmatic model creation for step input ---
% To keep the lesson self-contained, we show how to generate a simple Simulink
% model with a State-Space block and a Step block.
%
% Uncomment to create:
% local_build_simulink_model(A,B,[1 0],0,ts,u0);

% --------- local functions ---------
function Phi = local_phi_series(Z, m, terms)
    n = size(Z,1);
    Phi = eye(n)/factorial(m);
    Zpow = eye(n);
    for j = 1:terms-1
        Zpow = Zpow*Z;
        Phi = Phi + Zpow/factorial(j+m);
    end
end

function local_build_simulink_model(A,B,C,D,ts,amp)
    mdl = 'Chapter7_Lesson3_Simulink';
    if bdIsLoaded(mdl), close_system(mdl,0); end
    new_system(mdl);
    open_system(mdl);

    add_block('simulink/Sources/Step',[mdl '/Step'], 'Time', num2str(ts), ...
              'Before', '0', 'After', num2str(amp));
    add_block('simulink/Continuous/State-Space',[mdl '/StateSpace']);
    set_param([mdl '/StateSpace'], 'A', mat2str(A), 'B', mat2str(B), ...
              'C', mat2str(C), 'D', mat2str(D));
    add_block('simulink/Sinks/Scope',[mdl '/Scope']);

    set_param([mdl '/StateSpace'], 'Position', [200 80 350 150]);
    set_param([mdl '/Step'], 'Position', [50 90 100 120]);
    set_param([mdl '/Scope'], 'Position', [430 80 460 110]);

    add_line(mdl,'Step/1','StateSpace/1');
    add_line(mdl,'StateSpace/1','Scope/1');

    set_param(mdl,'StopTime','5');
    save_system(mdl);
end
